### PR TITLE
ZJIT: Put locals in stack map

### PR DESCRIPTION
**Pull request by alanwu-shopify-inc**
Prior to this, we have had the `zjit_jit_frame` (stack map) infra but have skipped over the portion of the stack that hosts local variables, spilling all locals to memory through EP eagerly before safepoints.

This commit puts locals into the stack map. The set of locals that are still eagerly written to memory are controled in HIR using the `spilled_locals` bit mask. Locals not spilled are described by the stack map. Eager spilling typically only happens before with-block sends for locals syntactically accessed inside the block. 

This still spills all locals prior to `invokebuiltin` because the design of `Primitive.cstmt` and `Primitive.cexpr` has each instance read all ruby locals for the C expression to use. This inefficiency seems okay for now because the core methods that use these primitives only have a handful of locals.

HIR has a tight grip on what locals are spilled now, but this stops short of using HIR `StoreField` for spilling for addressing <https://github.com/Shopify/ruby/issues/976>. Let's do that later.

---

Speed wise, it swings 1%~2% around neutral on ruby-bench headliners.
With some P-hacking you could claim there is a speedup:

```text
================================================================================
ZJIT Stats Comparison
================================================================================

  base (baseline):
    ruby 4.1.0dev (2026-09-01T21:21:55Z :detached: 837df2793b) +ZJIT +PRISM [aarch64-linux]
  branch:
    ruby 4.1.0dev (2026-09-01T21:58:16Z zjit-locals-in-sta.. c2a68ea8ff) +ZJIT +PRISM [aarch64-linux]
last_commit=reindent

BENCHMARK TIMINGS (lower is better)
--------------------------------------------------------------------------------
  activerecord:
    base                 avg:   0.256s  min:   0.252s    (baseline)
    branch               avg:   0.252s  min:   0.241s  ★    -1.4% (faster)
  chunky-png:
    base                 avg:   1.526s  min:   1.505s    (baseline)
    branch               avg:   1.466s  min:   1.434s  ★    -3.9% (faster)
  erubi-rails:
    base                 avg:   0.835s  min:   0.823s    (baseline)
    branch               avg:   0.778s  min:   0.768s  ★    -6.8% (faster)
  hexapdf:
    base                 avg:   2.395s  min:   2.089s    (baseline)
    branch               avg:   2.049s  min:   2.017s  ★   -14.5% (faster)
  liquid-c:
    base                 avg:   0.035s  min:   0.034s    (baseline)
    branch               avg:   0.034s  min:   0.031s  ★    -2.5% (faster)
  liquid-compile:
    base                 avg:   0.035s  min:   0.033s    (baseline)
    branch               avg:   0.033s  min:   0.031s  ★    -6.5% (faster)
  liquid-il:
    base                 avg:   0.381s  min:   0.378s    (baseline)
    branch               avg:   0.356s  min:   0.353s  ★    -6.5% (faster)
  liquid-render:
    base                 avg:   0.109s  min:   0.107s    (baseline)
    branch               avg:   0.105s  min:   0.102s  ★    -3.4% (faster)
  lobsters:
    base                 avg:   0.524s  min:   0.494s  ★ (baseline)
    branch               avg:   0.553s  min:   0.521s       +5.4% (slower)
  mail:
    base                 avg:   0.075s  min:   0.072s    (baseline)
    branch               avg:   0.072s  min:   0.070s  ★    -3.4% (faster)
  psych-load:
    base                 avg:   2.033s  min:   2.002s    (baseline)
    branch               avg:   1.984s  min:   1.957s  ★    -2.4% (faster)
  railsbench:
    base                 avg:   1.043s  min:   0.924s    (baseline)
    branch               avg:   0.903s  min:   0.878s  ★   -13.5% (faster)
  rubocop:
    base                 avg:   0.157s  min:   0.143s    (baseline)
    branch               avg:   0.138s  min:   0.129s  ★   -12.0% (faster)
  ruby-lsp:
    base                 avg:   0.134s  min:   0.131s    (baseline)
    branch               avg:   0.130s  min:   0.128s  ★    -2.8% (faster)
  sequel:
    base                 avg:   0.032s  min:   0.031s    (baseline)
    branch               avg:   0.031s  min:   0.030s  ★    -3.2% (faster)
  shipit:
    base                 avg:   0.909s  min:   0.874s    (baseline)
    branch               avg:   0.858s  min:   0.847s  ★    -5.6% (faster)
```

(lobsters is +/-5% on my setup. On an isolated run:

```
master: ruby 4.1.0dev (2026-09-02T10:02:32Z gh/master 76ce036fda) +ZJIT +PRISM [x86_64-linux]
+patch: ruby 4.1.0dev (2026-09-02T15:39:03Z zjit-locals-in-sta.. 921f321c80) +ZJIT +PRISM [x86_64-linux]

--------  -------------  ------------  --------------  -------------
bench       master (ms)   +patch (ms)  +patch 1st itr  master/+patch
lobsters  1002.6 ± 4.8%  995.4 ± 5.3%           1.024          1.007
--------  -------------  ------------  --------------  -------------
```

)

As expected, `vm_write_locals_count` is vastly reduced, but more time is spent compiling and more `zjit_alloc_bytes`:

```
SUMMARY COUNTERS (showing differences > 5.0%)
--------------------------------------------------------------------------------
  activerecord:
    vm_write_locals_count                  88,608,723 →      6,719,920 ▼  -92.4%
    gc_time                                       7ms →           11ms ▲  +52.7%
    profile_time                                  2ms →            3ms ▲  +10.7%
    compile_side_exit_time                       16ms →           17ms ▲  +10.4%
    zjit_alloc_bytes                        6,793,387 →      7,356,419 ▲   +8.3%
    invalidation_time                             0ms →            0ms ▲   +7.9%
    side_exit_size                          3,684,436 →      3,958,652 ▲   +7.4%
  chunky-png:
    vm_write_locals_count                 267,667,406 →        695,073 ▼  -99.7%
    gc_time                                       4ms →            3ms ▼  -23.8%
    vm_write_to_parent_iseq_local_count         5,064 →          4,269 ▼  -15.7%
    profile_time                                  0ms →            0ms ▼   -8.7%
    code_region_bytes                       1,245,184 →      1,146,880 ▼   -7.9%
    compile_time                                 52ms →           49ms ▼   -6.8%
    compile_lir_time                             35ms →           33ms ▼   -6.7%
    compile_side_exit_time                        2ms →            2ms ▼   -6.1%
    total_mem_bytes                         2,702,020 →      2,556,684 ▼   -5.4%
  erubi-rails:
    vm_write_locals_count                 244,056,512 →      7,314,362 ▼  -97.0%
    invalidation_time                             2ms →            2ms ▼  -13.0%
    profile_time                                  2ms →            2ms ▼   -5.3%
  hexapdf:
    vm_write_locals_count                 467,627,180 →      8,842,399 ▼  -98.1%
    invalidation_time                             0ms →            0ms ▲  +15.1%
    profile_time                                  3ms →            3ms ▲  +14.3%
    compile_side_exit_time                       15ms →           17ms ▲  +10.2%
    compile_lir_time                            219ms →          240ms ▲   +9.4%
    compile_time                                331ms →          361ms ▲   +9.2%
  liquid-c:
    vm_write_locals_count                   4,878,690 →        273,288 ▼  -94.4%
    invalidation_time                             0ms →            0ms ▲  +26.1%
    profile_time                                  0ms →            0ms ▲   +8.2%
  liquid-compile:
    vm_write_locals_count                   6,178,756 →        164,542 ▼  -97.3%
    vm_write_to_parent_iseq_local_count         1,231 →            651 ▼  -47.1%
    gc_time                                       3ms →            2ms ▼   -8.9%
    invalidation_time                             0ms →            0ms ▲   +6.1%
  liquid-il:
    vm_write_locals_count                 109,750,786 →      2,022,929 ▼  -98.2%
    invalidation_time                             0ms →            0ms ▼  -25.9%
    profile_time                                  1ms →            0ms ▼  -24.3%
    compile_lir_time                            101ms →           80ms ▼  -20.9%
    compile_time                                137ms →          108ms ▼  -20.6%
    gc_time                                       6ms →            5ms ▼  -16.4%
    compile_side_exit_time                        8ms →            7ms ▼  -11.8%
    code_region_bytes                       3,055,616 →      2,838,528 ▼   -7.1%
  liquid-render:
    vm_write_locals_count                  28,005,555 →      3,167,147 ▼  -88.7%
    invalidation_time                             0ms →            0ms ▼  -11.9%
    gc_time                                       3ms →            3ms ▼  -11.6%
  lobsters:
    vm_write_locals_count                 116,832,135 →      7,367,901 ▼  -93.7%
    compile_lir_time                          1,468ms →        1,943ms ▲  +32.3%
    compile_time                              2,198ms →        2,808ms ▲  +27.7%
    profile_time                                 28ms →           35ms ▲  +22.8%
    compile_side_exit_time                       96ms →          115ms ▲  +19.6%
    invalidation_time                             4ms →            5ms ▲  +15.3%
    side_exit_size                         19,792,928 →     21,306,364 ▲   +7.6%
    zjit_alloc_bytes                       35,517,237 →     38,068,798 ▲   +7.2%
    gc_time                                      52ms →           56ms ▲   +6.7%
  mail:
    vm_write_locals_count                  10,678,439 →        400,579 ▼  -96.2%
    invalidation_time                             0ms →            0ms ▼  -10.0%
    gc_time                                       7ms →            6ms ▼   -8.4%
    profile_time                                  1ms →            1ms ▼   -7.0%
    zjit_alloc_bytes                        3,528,012 →      3,744,900 ▲   +6.1%
  psych-load:
    vm_write_locals_count                 487,355,135 →      1,855,923 ▼  -99.6%
    gc_time                                      12ms →            2ms ▼  -81.5%
    vm_write_to_parent_iseq_local_count         1,113 →            823 ▼  -26.1%
    invalidation_time                             0ms →            0ms ▼  -10.2%
    profile_time                                  0ms →            0ms ▼   -6.9%
  railsbench:
    vm_write_locals_count                 214,824,192 →     14,087,887 ▼  -93.4%
    zjit_alloc_bytes                       13,593,580 →     14,583,921 ▲   +7.3%
    compile_side_exit_time                       33ms →           35ms ▲   +5.5%
    side_exit_size                          7,118,148 →      7,495,312 ▲   +5.3%
  rubocop:
    vm_write_locals_count                  37,157,888 →      1,989,948 ▼  -94.6%
    gc_time                                      24ms →           20ms ▼  -19.4%
    invalidation_time                             2ms →            2ms ▼  -12.8%
    side_exit_size                         15,714,228 →     17,255,396 ▲   +9.8%
    zjit_alloc_bytes                       22,570,842 →     24,626,190 ▲   +9.1%
    profile_time                                 14ms →           13ms ▼   -7.7%
    compile_side_exit_time                       73ms →           78ms ▲   +6.4%
  ruby-lsp:
    vm_write_locals_count                  24,945,193 →      1,585,358 ▼  -93.6%
    compile_side_exit_time                        8ms →            9ms ▲   +5.0%
  sequel:
    vm_write_locals_count                  19,293,828 →      2,414,572 ▼  -87.5%
    gc_time                                       3ms →            3ms ▼   -8.8%
    zjit_alloc_bytes                        2,189,412 →      2,306,814 ▲   +5.4%
    compile_side_exit_time                        4ms →            4ms ▲   +5.3%
  shipit:
    vm_write_locals_count                 206,054,432 →     13,126,175 ▼  -93.6%
    zjit_alloc_bytes                       26,161,651 →     28,644,535 ▲   +9.5%
    side_exit_size                         15,537,512 →     16,846,256 ▲   +8.4%
    compile_side_exit_time                       73ms →           78ms ▲   +7.0%
    profile_time                                 21ms →           22ms ▲   +6.4%
    gc_time                                     140ms →          132ms ▼   -5.8%
    invalidation_time                             3ms →            3ms ▼   -5.3%
```

These are my ruby/ruby-bench run json files you can use with [misc/zjit_diff.rb](https://github.com/ruby/ruby-bench/blob/main/misc/zjit_diff.rb):
[run_1.json](https://github.com/user-attachments/files/31760739/run_1.json)
[run_2.json](https://github.com/user-attachments/files/31760742/run_2.json)
[run_3.json](https://github.com/user-attachments/files/31760744/run_3.json)
[run_with_stats.json](https://github.com/user-attachments/files/31760745/run_with_stats.json)

### CODE REVIEW

File: vm.c [hunk 1]
```diff
@@ -1146,7 +1146,10 @@ vm_make_env_each(const rb_execution_context_t * const ec, rb_control_frame_t *co
     if (VM_FRAME_RUBYFRAME_P(cfp) &&
         !rbimpl_atomic_load(&ISEQ_BODY(iseq)->jit_ep_escape_recorded, RBIMPL_ATOMIC_RELAXED)) {
         if (rb_yjit_enabled_p) rb_yjit_invalidate_ep_is_bp(iseq);
-        if (rb_zjit_enabled_p) rb_zjit_invalidate_no_ep_escape(iseq);
+        if (rb_zjit_enabled_p) {
+            rb_zjit_invalidate_no_ep_escape(iseq);
+            rb_zjit_spill_frame(cfp);
```
Comment (k0kubun) [inline vm.c]: I think this deserves a comment about why we're spilling not only locals but also stack operands.

Technically, I believe we could skip materializing stack operands and an escaped Binding would still work properly. The reason why it's not doing that seems to be because the current stack map doesn't distinguish stack operands and locals, so we can't simply skip stack operand spills to minimize the number of memory writes, but it seems like something we could change if we want to (e.g. by adding a new kind of stack map entry for locals).

I'm not proposing to overcomplicate the implementation for the escaped path here though. I just wanted a comment for a non-obvious choice of the design.

File: zjit/src/hir.rs [hunk 2]
```diff
@@ -8066,7 +8097,12 @@ pub struct FrameState {
     pub pc: *const VALUE,
 
     stack: Vec<InsnId>,
+
+    /// Values for local variables in `rb_iseq_constant_body::local_table` order.
     locals: Vec<InsnId>,
+    /// The subset of locals that must be in EP memory prior to a safepoint.
+    /// Predominantly used for with-block sends for locals syntactically accessed within the block.
+    spilled_locals: LocalSet,
```
Comment (k0kubun) [inline zjit/src/hir.rs:8107]: I'm glad to see that we can finally do this and it's working.

However, the name of this field confused me. `spilled_locals` sounds like "locals that are already spilled" (well, sorry if it's not how English works) as if they should NOT be spilled anymore, which is the opposite of what needs to happen for them. I'd say it's more like `locals_to_spill`, `spill_locals`, or `block_accessed_locals`. WDYT?

**Comment by k0kubun** [comment]
> fix https://github.com/Shopify/ruby/issues/976

looks like the description will let this PR close the issue. you might want to change the wording a little to work around it.

File: zjit/src/codegen.rs [hunk 3]
```diff
@@ -3494,31 +3500,42 @@ fn build_stack_map(jit: &JITState, function: &Function, state: &FrameState) -> V
...
+                stack.push(to_entry(insn_id));
+            }
+        }
 
         let Some(caller) = current_state.caller() else {
             break;
         };
-        stack.push(StackMapEntry::Skip(inline_frame_stack_gap(current_state.iseq)));
+        // Skip the callee's receiver slot below its local table.
+        // We currently never map out the stack for `invokeblock`, which doesn't
```
Comment (k0kubun) [inline zjit/src/codegen.rs]: Sorry, I couldn't follow what "We currently never map out the stack for `invokeblock`" means. Is that just saying we don't use stack maps for invokeblock and we eagerly spill stack slots?

File: zjit/src/codegen.rs [hunk 4]
```diff
@@ -3494,31 +3500,42 @@ fn build_stack_map(jit: &JITState, function: &Function, state: &FrameState) -> V
...
+        // Operand stack, top-down.
+        stack.extend(current_state.stack().rev().copied().map(to_entry));
+        // Frame environment data (me/cref, specval, flags) already lives in memory.
+        stack.push(StackMapEntry::Skip(VM_ENV_DATA_SIZE.to_usize()));
+        // Locals, top-down (local[L-1] .. local[0]).
+        let spilled_locals = current_state.spilled_locals();
+        for (idx, &insn_id) in current_state.locals().enumerate().rev() {
+            // Skip spilled slots which are already initialized on entry.
+            if spilled_locals.get(idx) {
+                stack.push(StackMapEntry::Skip(1));
```
Comment (k0kubun) [inline zjit/src/codegen.rs]: It might be nice to batch those skip entries to minimize the run-time overhead when they are adjacent to each other.

**Comment by k0kubun** [comment]
If you have encountered any difficulties that you needed to debug while creating this patch, it'd be nice to include those cases as regression tests in `jit_frame.rs` or `codegen_test.rs`. Is every corner case for local spills covered in existing tests?

**Review by k0kubun — APPROVED** [review]
looks great. I'm surprised to see that the diff of the local support can be this simple. I was worried that we'd need more unexpected hacks.

File: zjit/src/hir.rs (hunk 2 above)
Comment (XrXr) [reply zjit/src/hir.rs:8107]: I think it makes sense as is, looking at FrameState as declarative. The codegen part that actually does the stores for this can be thought of as changing the state to what the FrameState specifies. I feel when the actual stores move to HIR it'll feel even more declarative, since the state snapshot will follow all the stores.

I should rephrase the doc to adopt this declarative framing, though. "The subset of locals that are in EP memory."

Comment (k0kubun) [reply zjit/src/hir.rs:8107]: > I should rephrase the doc to adopt this declarative framing, though. "The subset of locals that are in EP memory."

The current "The subset of locals that must be in EP memory prior to a safepoint" seems clearer than "The subset of locals that are in EP memory" though. They aren't in EP memory when codegen starts reading the FrameState. It's more that they "need to be" in EP memory. I would probably phrase it like "The subset of locals that should be spilled into EP memory prior to a safepoint" (or "prior to a non-leaf call").

File: zjit/src/codegen.rs (hunk 4 above)
Comment (XrXr) [reply zjit/src/codegen.rs]: Good call! 97f7b4e9c9

File: zjit/src/codegen.rs (hunk 3 above)
Comment (XrXr) [reply zjit/src/codegen.rs]: It has to do with the inliner. I put 6785679e23 to explain it better.

File: vm.c (hunk 1 above)
Comment (XrXr) [reply vm.c]: Thanks, looking at this led me to a bug with this! Fix for that and comment about the design choice: 05fb09d798

File: zjit/src/hir.rs (hunk 2 above)
Comment (XrXr) [reply zjit/src/hir.rs:8107]: Ah, "The subset of locals that are in EP memory" is somewhat incorrect today. I see the issue. Maybe we name it `locals_to_spill` today and rename it once the spills are in HIR. Maybe we should get away from saying "spill" entirely because it's no really overflow anything. Maybe `mem_locals`, which sounds okay today and in the future.

**Comment by XrXr** [comment]
This is on hold for now because we want to build build fixes on top of the locals in stack map first to proof out the design.
